train.py

```diff
@@ -6,51 +8,300 @@
+        return state, metrics
+
+
+class Evaluater:
+    """A class to evaluate the model with, save and checkpoint loss metrics."""
+    def __init__(self, net, loss_fn, checkpoint_dir: str,
+            checkpoint_every_n: int, eval_every_n: int):
+        self._net_init = net.init
+        self._net_apply = net.apply
+        self._loss_fn = loss_fn
+        self.val_queue = []
+        self.best_state = None # save the state with lowest validation error in best state
+        self.lowest_val_loss = None
+        self._checkpoint_dir = checkpoint_dir
+        self._checkpoint_every_n = checkpoint_every_n
+        self._eval_every_n = eval_every_n
+        # load loss curve if metrics file exists in checkpoint_dir
+        metrics_path = os.path.join(self._checkpoint_dir, 'metrics.pkl')
+        best_state_path = os.path.join(self._checkpoint_dir, 
+                'best_state.pkl')
+        if os.path.exists(metrics_path):
+            # load metrics, if they have been saved before
+            logging.info('Loading metrics from %s', metrics_path)
+            with open(metrics_path, 'rb') as f:
+                self._metrics_dict = pickle.load(f)
+            self._loaded_metrics = True
+            # load best state and lowest loss
+            with open(best_state_path, 'rb') as f:
+                best_state_dict = pickle.load(f)
+                self.best_state = best_state_dict['state']
+                self.lowest_val_loss = best_state_dict['loss']
+        else:
+            self._loaded_metrics = False
+
+    @functools.partial(jax.jit, static_argnums=0)
+    def _evaluate_step(
+            self, state: dict, graphs: jraph.GraphsTuple) -> float:
+        """Calculate the mean loss for a batch of graphs."""
+        mean_loss = self._loss_fn(state['params'], graphs, self._net_apply)
+        return mean_loss
+
+    def evaluate_split(
+            self, 
+            state: dict,
+            graphs: Sequence[jraph.GraphsTuple],
+            batch_size: int) -> float:
+        """Return mean loss for all graphs in graphs."""
+        
+        reader = DataReader(data=graphs, 
+            batch_size=batch_size, repeat=False)
+        
+        loss_list = [self._evaluate_step(state, batch) for batch in reader]
+        return np.mean(loss_list)
+
+    def evaluate_model(
+            self, 
+            state: Dict,
+            datasets: Dict[str, Iterable[jraph.GraphsTuple]],
+            splits: Iterable[str]) -> Dict[str, float]:
+        """Return mean loss for every split in splits."""
+        eval_loss = {}
+        for split in splits:
+            eval_loss[split] = self.evaluate_split(state, 
+                    datasets[split].data, datasets[split].batch_size)
+            if split=='validation':
+                if self.best_state is None or eval_loss[split] < self.lowest_val_loss:
+                    self.best_state = state.copy()
+                    self.lowest_val_loss = eval_loss[split]
+        return eval_loss
```

Comment:
The name eval loss is a bit confusing here since it looks like it's actually a dict that might contain the loss of the train/val split or did I misunderstand something here?